File: python/exercise2/DoublyLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.previous = None
        self.next = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

        # TO check weather list is empty or not?

    def is_empty(self):
        if self.head is None:
            return True
        else:
            return False

    def add_node(self, data):
        #  Creating New Node
        newNode = Node(data)

        # Checking where the head and tail pointer is pointing.
        # if first node is added to the list then below condition should be true.
        if self.head is None:
            self.head = newNode
            newNode.previous = newNode.next = None

        else:
            # Addding new node at the last
            self.tail.next = newNode
            newNode.previous = self.tail
            newNode.next = None
        self.tail = newNode

    def display_list(self):
        # Creating temp pointer to travers
        temp = self.head
        if temp is not None:
            while temp is not None:
                print(temp.data, end=" -> ")
                temp = temp.next
            else:
                print("NULL")
        else:
            print("List does not have any nodes")

    def clone(self):
        new_list = DoublyLinkedList()
        current = self.head
        while current is not None:
            new_list.add_node(current.data)
            current = current.next
        return new_list

    def concatenate(self, list2):
        if self.is_empty():
            self.head = list2.head
            self.tail = list2.tail
        elif not list2.is_empty():
            self.tail.next = list2.head
            list2.head.previous = self.tail
            self.tail = list2.tail


if __name__ == "__main__":
    L1 = DoublyLinkedList()
    L1.add_node("MSP")
    L1.add_node("ATL")
    L1.add_node("BOS")
    L1.display_list()

    L2 = DoublyLinkedList()
    L2.add_node("LAX")
    L2.add_node("ABC")
    L2.add_node("XYZ")
    L2.add_node("JFK")
    L2.display_list()

    L = L1.clone()
    L.concatenate(L2)
    L1.display_list()
    L2.display_list()
    L.display_list()

File: java/exercise2/DoublyLinkedList.java
package exercise2;

public class DoublyLinkedList<E> {
  private static class Node<E> {
    private E element;
    private Node<E> prev;
    private Node<E> next;

    public Node(E e, Node<E> p, Node<E> n) {
      element = e;
      prev = p;
      next = n;
    }

    public E getElement() {
      return element;
    }

    public Node<E> getPrev() {
      return prev;
    }

    public Node<E> getNext() {
      return next;
    }

    public void setPrev(Node<E> p) {
      prev = p;
    }

    public void setNext(Node<E> n) {
      next = n;
    }
  }

  private Node<E> header;
  private Node<E> trailer;
  private int size = 0;

  public DoublyLinkedList() {
    header = new Node<>(null, null, null);
    trailer = new Node<>(null, header, null);
    header.setNext(trailer);
  }

  public int size() {
    return size;
  }

  public boolean isEmpty() {
    return size == 0;
  }

  public E first() {
    if (isEmpty())
      return null;
    return header.getNext().getElement();
  }

  public E last() {
    if (isEmpty())
      return null;
    return trailer.getPrev().getElement();
  }

  public void addFirst(E e) {
    addBetween(e, header, header.getNext());
  }

  public void addLast(E e) {
    addBetween(e, trailer.getPrev(), trailer);
  }

  public E removeFirst() {
    if (isEmpty())
      return null;
    return remove(header.getNext());
  }

  public E removeLast() {
    if (isEmpty())
      return null;
    return remove(trailer.getPrev());
  }

  private void addBetween(E e, Node<E> predecessor, Node<E> successor) {
    Node<E> newest = new Node<>(e, predecessor, successor);
    predecessor.setNext(newest);
    successor.setPrev(newest);
    size++;
  }

  private E remove(Node<E> node) {
    Node<E> predecessor = node.getPrev();
    Node<E> successor = node.getNext();
    predecessor.setNext(successor);
    successor.setPrev(predecessor);
    size--;
    return node.getElement();
  }

  public String toString() {
    StringBuilder sb = new StringBuilder("(");
    Node<E> walk = header.getNext();
    while (walk != trailer) {
      sb.append(walk.getElement());
      walk = walk.getNext();
      if (walk != trailer)
        sb.append(", ");
    }
    sb.append(")");
    return sb.toString();
  }

  public static void main(String[] args) {
    DoublyLinkedList<String> L1 = new DoublyLinkedList<String>();
    L1.addFirst("MSP");
    L1.addLast("ATL");
    L1.addLast("BOS");
    //
    L1.addFirst("LAX");
    System.out.println(L1);

    DoublyLinkedList<String> L2 = new DoublyLinkedList<String>();
    L2.addFirst("YYZ");
    L2.addLast("YVR");
    System.out.println(L2);

    DoublyLinkedList<String> L = L1.clone();
    L.concatenate(L2);
    System.out.println(L1);
    System.out.println(L2);
    System.out.println(L);
  }

  public DoublyLinkedList<E> clone() {
    DoublyLinkedList<E> cloneList = new DoublyLinkedList<>();
    Node<E> current = this.header.getNext();
    while (current != this.trailer) {
      cloneList.addLast(current.getElement());
      current = current.getNext();
    }
    return cloneList;
  }

  private void concatenate(DoublyLinkedList<String> list2) {
    // TODO Auto-generated method stub
    trailer.prev.setNext((Node<E>) list2.header.next);
    list2.header.next.setPrev((Node<String>) trailer.getPrev());
    trailer = (Node<E>) list2.trailer;
  }
}
